File: examples/todo_app/java_todo/src/main/java/io/jamharah/todo/TodoStore.java
/* SPDX-License-Identifier: Apache-2.0 */
package io.jamharah.todo;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.function.Consumer;

/**
 * SQLite persistence for todo_app (port of {@code nim_todo/todo_store.nim} / {@code
 * rust_todo/src/todo_store.rs}).
 */
public final class TodoStore implements AutoCloseable {

  private static final int TITLE_MAX = 255;
  private static final int DESC_MAX = 4096;
  private static final int NAME_MAX = 64;
  private static final int COLOR_MAX = 32;

  private static final String SCHEMA_DDL_V3 =
      "CREATE TABLE IF NOT EXISTS categories (\n"
          + "  id INTEGER PRIMARY KEY AUTOINCREMENT,\n"
          + "  name TEXT NOT NULL UNIQUE CHECK(length(name) BETWEEN 1 AND 64),\n"
          + "  color TEXT NOT NULL DEFAULT '#6366f1' CHECK(length(color) BETWEEN 1 AND 32)\n"
          + ");\n"
          + "CREATE TABLE IF NOT EXISTS todos (\n"
          + "  id INTEGER PRIMARY KEY AUTOINCREMENT,\n"
          + "  title TEXT NOT NULL CHECK(length(title) BETWEEN 1 AND 255),\n"
          + "  description TEXT NOT NULL DEFAULT '' CHECK(length(description) <= 4096),\n"
          + "  status TEXT NOT NULL DEFAULT 'pending'\n"
          + "    CHECK(status IN ('pending','in_progress','done','cancelled','archived')),\n"
          + "  priority TEXT NOT NULL DEFAULT 'medium'\n"
          + "    CHECK(priority IN ('low','medium','high','urgent')),\n"
          + "  category_id INTEGER REFERENCES categories(id) ON DELETE SET NULL,\n"
          + "  parent_id INTEGER REFERENCES todos(id) ON DELETE CASCADE,\n"
          + "  due_date INTEGER,\n"
          + "  position INTEGER NOT NULL DEFAULT 0,\n"
          + "  created_at INTEGER NOT NULL DEFAULT (unixepoch()),\n"
          + "  updated_at INTEGER NOT NULL DEFAULT (unixepoch())\n"
          + ");\n"
          + "CREATE INDEX IF NOT EXISTS idx_todos_status ON todos(status);\n"
          + "CREATE INDEX IF NOT EXISTS idx_todos_priority ON todos(priority);\n"
          + "CREATE INDEX IF NOT EXISTS idx_todos_category ON todos(category_id);\n";

  private static final String MIGRATE_V2_TO_V3 =
      "CREATE TABLE todos_new (\n"
          + "  id INTEGER PRIMARY KEY AUTOINCREMENT,\n"
          + "  title TEXT NOT NULL CHECK(length(title) BETWEEN 1 AND 255),\n"
          + "  description TEXT NOT NULL DEFAULT '' CHECK(length(description) <= 4096),\n"
          + "  status TEXT NOT NULL DEFAULT 'pending'\n"
          + "    CHECK(status IN ('pending','in_progress','done','cancelled','archived')),\n"
          + "  priority TEXT NOT NULL DEFAULT 'medium'\n"
          + "    CHECK(priority IN ('low','medium','high','urgent')),\n"
          + "  category_id INTEGER REFERENCES categories(id) ON DELETE SET NULL,\n"
          + "  parent_id INTEGER REFERENCES todos(id) ON DELETE CASCADE,\n"
          + "  due_date INTEGER,\n"
          + "  position INTEGER NOT NULL DEFAULT 0,\n"
          + "  created_at INTEGER NOT NULL DEFAULT (unixepoch()),\n"
          + "  updated_at INTEGER NOT NULL DEFAULT (unixepoch())\n"
          + ");\n"
          + "INSERT INTO todos_new(id, title, description, status, priority, category_id, parent_id, due_date,\n"
          + " position, created_at, updated_at)\n"
          + "SELECT id, title, description, status, priority, category_id, parent_id, due_date, position,\n"
          + "created_at, updated_at FROM todos;\n"
          + "DROP TABLE todos;\n"
          + "ALTER TABLE todos_new RENAME TO todos;\n"
          + "CREATE INDEX IF NOT EXISTS idx_todos_status ON todos(status);\n"
          + "CREATE INDEX IF NOT EXISTS idx_todos_priority ON todos(priority);\n"
          + "CREATE INDEX IF NOT EXISTS idx_todos_category ON todos(category_id);\n"
          + "PRAGMA user_version = 3;\n";

  private final Connection conn;

  private TodoStore(Connection conn) {
    this.conn = conn;
  }

  private static int mapErr(SQLException e) {
    int c = e.getErrorCode();
    return c != 0 ? c : SqliteRc.SQLITE_ERROR;
  }

  private static int execSql(Connection db, String sql) {
    for (String part : sql.split(";")) {
      String p = part.trim();
      if (p.isEmpty()) {
        continue;
      }
      try (Statement st = db.createStatement()) {
        st.execute(p);
      } catch (SQLException e) {
        System.err.println("[todo_store] " + e.getMessage());
        return mapErr(e);
      }
    }
    return SqliteRc.SQLITE_OK;
  }

  private static int readUserVersion(Connection db) {
    try (Statement st = db.createStatement();
        ResultSet rs = st.executeQuery("PRAGMA user_version;")) {
      if (rs.next()) {
        return rs.getInt(1);
      }
    } catch (SQLException ignored) {
      // fall through
    }
    return -1;
  }

  private static int ensureSchema(Connection db) throws SQLException {
    int ver = readUserVersion(db);
    if (ver < 0) {
      return SqliteRc.SQLITE_ERROR;
    }
    if (execSql(db, "PRAGMA foreign_keys = ON;") != SqliteRc.SQLITE_OK) {
      return SqliteRc.SQLITE_ERROR;
    }
    execSql(db, "PRAGMA journal_mode=WAL;");
    if (ver == 1) {
      if (execSql(db, "DROP TABLE IF EXISTS todos;") != SqliteRc.SQLITE_OK) {
        return SqliteRc.SQLITE_ERROR;
      }
    }
    if (ver == 2) {
      return execSql(db, MIGRATE_V2_TO_V3);
    }
    if (execSql(db, SCHEMA_DDL_V3) != SqliteRc.SQLITE_OK) {
      return SqliteRc.SQLITE_ERROR;
    }
    ver = readUserVersion(db);
    if (ver < 3 && execSql(db, "PRAGMA user_version = 3;") != SqliteRc.SQLITE_OK) {
      return SqliteRc.SQLITE_ERROR;
    }
    return SqliteRc.SQLITE_OK;
  }

  /** Opens the store; empty {@code path} uses in-memory SQLite. */
  public static TodoStore storeOpen(String path) {
    String uri = path == null || path.isEmpty() ? ":memory:" : path;
    String url = "jdbc:sqlite:" + uri;
    try {
      Connection c = DriverManager.getConnection(url);
      try (Statement st = c.createStatement()) {
        st.execute("PRAGMA busy_timeout = 5000");
      }
      if (ensureSchema(c) != SqliteRc.SQLITE_OK) {
        c.close();
        return null;
      }
      return new TodoStore(c);
    } catch (SQLException e) {
      System.err.println("[todo_store] open failed: " + e.getMessage());
      return null;
    }
  }

  public static void storeClose(TodoStore s) {
    if (s != null) {
      s.close();
    }
  }

  @Override
  public void close() {
    try {
      conn.close();
    } catch (SQLException ignored) {
      // ignore
    }
  }

  private static int validateNameColor(String name, String color) {
    if (name == null || color == null || name.isEmpty() || color.isEmpty()) {
      return SqliteRc.SQLITE_MISUSE;
    }
    int nl = name.length();
    int cl = color.length();
    if (nl < 1 || nl > NAME_MAX || cl < 1 || cl > COLOR_MAX) {
      return SqliteRc.SQLITE_RANGE;
    }
    return SqliteRc.SQLITE_OK;
  }

  private static int validateTodoTexts(String title, String description, String status, String priority) {
    if (status == null || priority == null || status.isEmpty() || priority.isEmpty()) {
      return SqliteRc.SQLITE_MISUSE;
    }
    if (title == null || description == null) {
      return SqliteRc.SQLITE_MISUSE;
    }
    int tl = title.length();
    int dl = description.length();
    if (tl < 1 || tl > TITLE_MAX || dl > DESC_MAX) {
      return SqliteRc.SQLITE_RANGE;
    }
    if (!status.equals("pending")
        && !status.equals("in_progress")
        && !status.equals("done")
        && !status.equals("cancelled")
        && !status.equals("archived")) {
      return SqliteRc.SQLITE_CONSTRAINT;
    }
    if (!priority.equals("low")
        && !priority.equals("medium")
        && !priority.equals("high")
        && !priority.equals("urgent")) {
      return SqliteRc.SQLITE_CONSTRAINT;
    }
    return SqliteRc.SQLITE_OK;
  }

  public int storeCategoriesForeach(Consumer<TodoStoreCategoryRow> fn) {
    if (fn == null) {
      return SqliteRc.SQLITE_MISUSE;
    }
    String q = "SELECT id, name, color FROM categories ORDER BY name COLLATE NOCASE ASC;";
    try (PreparedStatement ps = conn.prepareStatement(q);
        ResultSet rs = ps.executeQuery()) {
      while (rs.next()) {
        fn.accept(
            new TodoStoreCategoryRow(rs.getLong(1), rs.getString(2), rs.getString(3)));
      }
      return SqliteRc.SQLITE_OK;
    } catch (SQLException e) {
      return mapErr(e);
    }
  }

  public int storeCategoryInsert(String name, String color, long[] outId) {
    if (outId == null || outId.length < 1) {
      return SqliteRc.SQLITE_MISUSE;
    }
    int v = validateNameColor(name, color);
    if (v != SqliteRc.SQLITE_OK) {
      return v;
    }
    String q = "INSERT INTO categories(name, color) VALUES(?, ?);";
    try (PreparedStatement ps = conn.prepareStatement(q, Statement.RETURN_GENERATED_KEYS)) {
      ps.setString(1, name);
      ps.setString(2, color);
      ps.executeUpdate();
      try (ResultSet keys = ps.getGeneratedKeys()) {
        if (keys.next()) {
          outId[0] = keys.getLong(1);
        }
      }
      if (outId[0] <= 0) {
        return SqliteRc.SQLITE_ERROR;
      }
      return SqliteRc.SQLITE_OK;
    } catch (SQLException e) {
      return mapErr(e);
    }
  }

  public int storeCategoryDeleteById(long id) {
    String q = "DELETE FROM categories WHERE id = ?;";
    try (PreparedStatement ps = conn.prepareStatement(q)) {
      ps.setLong(1, id);
      return ps.executeUpdate() > 0 ? SqliteRc.SQLITE_OK : SqliteRc.SQLITE_ERROR;
    } catch (SQLException e) {
      return mapErr(e);
    }
  }

  public int storeTodosForeach(Consumer<TodoStoreTodoRow> fn) {
    if (fn == null) {
      return SqliteRc.SQLITE_MISUSE;
    }
    String q =
        "SELECT id, title, description, status, priority, category_id, parent_id, due_date,"
            + " position, created_at, updated_at FROM todos ORDER BY position ASC, id ASC;";
    try (PreparedStatement ps = conn.prepareStatement(q);
        ResultSet rs = ps.executeQuery()) {
      while (rs.next()) {
        long cat = rs.getLong(6);
        boolean catNull = rs.wasNull();
        long par = rs.getLong(7);
        boolean parNull = rs.wasNull();
        long due = rs.getLong(8);
        boolean dueNull = rs.wasNull();
        fn.accept(
            new TodoStoreTodoRow(
                rs.getLong(1),
                rs.getString(2),
                rs.getString(3),
                rs.getString(4),
                rs.getString(5),
                catNull,
                cat,
                parNull,
                par,
                dueNull,
                due,
                rs.getInt(9),
                rs.getLong(10),
                rs.getLong(11)));
      }
      return SqliteRc.SQLITE_OK;
    } catch (SQLException e) {
      return mapErr(e);
    }
  }

  public int storeTodoInsert(
      String title,
      String description,
      String status,
      String priority,
      boolean categoryIdSet,
      long categoryId,
      boolean parentIdSet,
      long parentId,
      boolean dueDateSet,
      long dueDate,
      long[] outId) {
    if (outId == null || outId.length < 1) {
      return SqliteRc.SQLITE_MISUSE;
    }
    int v = validateTodoTexts(title, description, status, priority);
    if (v != SqliteRc.SQLITE_OK) {
      return v;
    }
    String q =
        "INSERT INTO todos(title, description, status, priority, category_id, parent_id, due_date, position)"
            + " VALUES(?,?,?,?,?,?,?, (SELECT COALESCE(MAX(position), -1) + 1 FROM todos));";
    try (PreparedStatement ps = conn.prepareStatement(q, Statement.RETURN_GENERATED_KEYS)) {
      ps.setString(1, title);
      ps.setString(2, description);
      ps.setString(3, status);
      ps.setString(4, priority);
      if (categoryIdSet && categoryId > 0) {
        ps.setLong(5, categoryId);
      } else {
        ps.setNull(5, java.sql.Types.INTEGER);
      }
      if (parentIdSet && parentId > 0) {
        ps.setLong(6, parentId);
      } else {
        ps.setNull(6, java.sql.Types.INTEGER);
      }
      if (dueDateSet) {
        ps.setLong(7, dueDate);
      } else {
        ps.setNull(7, java.sql.Types.INTEGER);
      }
      ps.executeUpdate();
      outId[0] = 0;
      try (ResultSet keys = ps.getGeneratedKeys()) {
        if (keys.next()) {
          outId[0] = keys.getLong(1);
        }
      }
      if (outId[0] <= 0) {
        try (Statement st = conn.createStatement();
            ResultSet rs = st.executeQuery("SELECT last_insert_rowid();")) {
          if (rs.next()) {
            outId[0] = rs.getLong(1);
          }
        }
      }
      if (outId[0] <= 0) {
        return SqliteRc.SQLITE_ERROR;
      }
      return SqliteRc.SQLITE_OK;
    } catch (SQLException e) {
      return mapErr(e);
    }
  }

  public int storeTodoUpdateFull(
      long id,
      String title,
      String description,
      String status,
      String priority,
      boolean categoryIdSet,
      long categoryId,
      boolean parentIdSet,
      long parentId,
      boolean dueDateSet,
      long dueDate,
      int position) {
    int v = validateTodoTexts(title, description, status, priority);
    if (v != SqliteRc.SQLITE_OK) {
      return v;
    }
    String q =
        "UPDATE todos SET title=?, description=?, status=?, priority=?, category_id=?, parent_id=?"
            + ", due_date=?, position=?, updated_at=unixepoch() WHERE id=?;";
    try (PreparedStatement ps = conn.prepareStatement(q)) {
      ps.setString(1, title);
      ps.setString(2, description);
      ps.setString(3, status);
      ps.setString(4, priority);
      if (categoryIdSet && categoryId > 0) {
        ps.setLong(5, categoryId);
      } else {
        ps.setNull(5, java.sql.Types.INTEGER);
      }
      if (parentIdSet && parentId > 0) {
        ps.setLong(6, parentId);
      } else {
        ps.setNull(6, java.sql.Types.INTEGER);
      }
      if (dueDateSet) {
        ps.setLong(7, dueDate);
      } else {
        ps.setNull(7, java.sql.Types.INTEGER);
      }
      ps.setInt(8, position);
      ps.setLong(9, id);
      return ps.executeUpdate() > 0 ? SqliteRc.SQLITE_OK : SqliteRc.SQLITE_ERROR;
    } catch (SQLException e) {
      return mapErr(e);
    }
  }

  public int storeTodoDeleteById(long id) {
    String q = "DELETE FROM todos WHERE id = ?;";
    try (PreparedStatement ps = conn.prepareStatement(q)) {
      ps.setLong(1, id);
      return ps.executeUpdate() > 0 ? SqliteRc.SQLITE_OK : SqliteRc.SQLITE_ERROR;
    } catch (SQLException e) {
      return mapErr(e);
    }
  }
}

File: examples/todo_app/csharp_todo/TodoStore.cs
// SQLite persistence for todo_app (port of c_todo/src/todo_store.c / nim_todo/todo_store.nim).
// SPDX-License-Identifier: Apache-2.0

using Microsoft.Data.Sqlite;

namespace Jamharah.Todo;

public static class SqliteRc
{
    public const int Ok = 0;
    public const int Error = 1;
    public const int Misuse = 21;
    public const int Range = 25;
    public const int Constraint = 19;
}

public sealed record TodoStoreCategoryRow(long Id, string Name, string Color);

public sealed record TodoStoreTodoRow(
    long Id,
    string Title,
    string Description,
    string Status,
    string Priority,
    bool CategoryIdIsNull,
    long CategoryId,
    bool ParentIdIsNull,
    long ParentId,
    bool DueDateIsNull,
    long DueDate,
    int Position,
    long CreatedAt,
    long UpdatedAt);

public sealed class TodoStore
{
    internal SqliteConnection Conn { get; }

    internal TodoStore(SqliteConnection conn) => Conn = conn;
}

public static class TodoStoreApi
{
    private const int TitleMax = 255;
    private const int DescMax = 4096;
    private const int NameMax = 64;
    private const int ColorMax = 32;

    private const string SchemaDdlV3 = """
        CREATE TABLE IF NOT EXISTS categories (
          id INTEGER PRIMARY KEY AUTOINCREMENT,
          name TEXT NOT NULL UNIQUE CHECK(length(name) BETWEEN 1 AND 64),
          color TEXT NOT NULL DEFAULT '#6366f1' CHECK(length(color) BETWEEN 1 AND 32)
        );
        CREATE TABLE IF NOT EXISTS todos (
          id INTEGER PRIMARY KEY AUTOINCREMENT,
          title TEXT NOT NULL CHECK(length(title) BETWEEN 1 AND 255),
          description TEXT NOT NULL DEFAULT '' CHECK(length(description) <= 4096),
          status TEXT NOT NULL DEFAULT 'pending'
            CHECK(status IN ('pending','in_progress','done','cancelled','archived')),
          priority TEXT NOT NULL DEFAULT 'medium'
            CHECK(priority IN ('low','medium','high','urgent')),
          category_id INTEGER REFERENCES categories(id) ON DELETE SET NULL,
          parent_id INTEGER REFERENCES todos(id) ON DELETE CASCADE,
          due_date INTEGER,
          position INTEGER NOT NULL DEFAULT 0,
          created_at INTEGER NOT NULL DEFAULT (unixepoch()),
          updated_at INTEGER NOT NULL DEFAULT (unixepoch())
        );
        CREATE INDEX IF NOT EXISTS idx_todos_status ON todos(status);
        CREATE INDEX IF NOT EXISTS idx_todos_priority ON todos(priority);
        CREATE INDEX IF NOT EXISTS idx_todos_category ON todos(category_id);
        """;

    private const string MigrateV2ToV3 = """
        CREATE TABLE todos_new (
          id INTEGER PRIMARY KEY AUTOINCREMENT,
          title TEXT NOT NULL CHECK(length(title) BETWEEN 1 AND 255),
          description TEXT NOT NULL DEFAULT '' CHECK(length(description) <= 4096),
          status TEXT NOT NULL DEFAULT 'pending'
            CHECK(status IN ('pending','in_progress','done','cancelled','archived')),
          priority TEXT NOT NULL DEFAULT 'medium'
            CHECK(priority IN ('low','medium','high','urgent')),
          category_id INTEGER REFERENCES categories(id) ON DELETE SET NULL,
          parent_id INTEGER REFERENCES todos(id) ON DELETE CASCADE,
          due_date INTEGER,
          position INTEGER NOT NULL DEFAULT 0,
          created_at INTEGER NOT NULL DEFAULT (unixepoch()),
          updated_at INTEGER NOT NULL DEFAULT (unixepoch())
        );
        INSERT INTO todos_new(id, title, description, status, priority, category_id, parent_id, due_date,
         position, created_at, updated_at)
        SELECT id, title, description, status, priority, category_id, parent_id, due_date, position,
        created_at, updated_at FROM todos;
        DROP TABLE todos;
        ALTER TABLE todos_new RENAME TO todos;
        CREATE INDEX IF NOT EXISTS idx_todos_status ON todos(status);
        CREATE INDEX IF NOT EXISTS idx_todos_priority ON todos(priority);
        CREATE INDEX IF NOT EXISTS idx_todos_category ON todos(category_id);
        PRAGMA user_version = 3;
        """;

    public static void StoreClose(TodoStore? s)
    {
        if (s is null)
            return;
        s.Conn.Close();
        s.Conn.Dispose();
    }

    public static TodoStore? StoreOpen(string path)
    {
        var uri = string.IsNullOrEmpty(path) ? ":memory:" : path;
        try
        {
            var conn = new SqliteConnection($"Data Source={uri}");
            conn.Open();
            using (var cmd = conn.CreateCommand())
            {
                cmd.CommandText = "PRAGMA busy_timeout = 5000;";
                cmd.ExecuteNonQuery();
            }
            if (EnsureSchema(conn) != SqliteRc.Ok)
            {
                conn.Close();
                conn.Dispose();
                return null;
            }
            return new TodoStore(conn);
        }
        catch (SqliteException)
        {
            Console.Error.WriteLine("[todo_store] open failed");
            return null;
        }
    }

    private static int ExecSql(SqliteConnection conn, string sql)
    {
        try
        {
            using var cmd = conn.CreateCommand();
            cmd.CommandText = sql;
            cmd.ExecuteNonQuery();
            return SqliteRc.Ok;
        }
        catch (SqliteException e)
        {
            Console.Error.WriteLine($"[todo_store] {e.Message}");
            return SqliteRc.Error;
        }
    }

    private static int ReadUserVersion(SqliteConnection conn)
    {
        try
        {
            using var cmd = conn.CreateCommand();
            cmd.CommandText = "PRAGMA user_version;";
            var v = cmd.ExecuteScalar();
            return v is long l ? (int)l : v is int i ? i : -1;
        }
        catch (SqliteException)
        {
            return -1;
        }
    }

    private static int EnsureSchema(SqliteConnection conn)
    {
        var ver = ReadUserVersion(conn);
        if (ver < 0)
            return SqliteRc.Error;

        if (ExecSql(conn, "PRAGMA foreign_keys = ON;") != SqliteRc.Ok)
            return SqliteRc.Error;
        ExecSql(conn, "PRAGMA journal_mode=WAL;");

        if (ver == 1)
        {
            if (ExecSql(conn, "DROP TABLE IF EXISTS todos;") != SqliteRc.Ok)
                return SqliteRc.Error;
        }

        if (ver == 2)
        {
            try
            {
                using var cmd = conn.CreateCommand();
                cmd.CommandText = MigrateV2ToV3;
                cmd.ExecuteNonQuery();
            }
            catch (SqliteException e)
            {
                Console.Error.WriteLine($"[todo_store] {e.Message}");
                return SqliteRc.Error;
            }
            return SqliteRc.Ok;
        }

        if (ExecSql(conn, SchemaDdlV3) != SqliteRc.Ok)
            return SqliteRc.Error;

        ver = ReadUserVersion(conn);
        if (ver < 3 && ExecSql(conn, "PRAGMA user_version = 3;") != SqliteRc.Ok)
            return SqliteRc.Error;
        return SqliteRc.Ok;
    }

    private static int ValidateNameColor(string name, string color)
    {
        if (string.IsNullOrEmpty(name) || string.IsNullOrEmpty(color))
            return SqliteRc.Misuse;
        if (name.Length is < 1 or > NameMax || color.Length is < 1 or > ColorMax)
            return SqliteRc.Range;
        return SqliteRc.Ok;
    }

    private static int ValidateTodoTexts(string title, string description, string status, string priority)
    {
        if (string.IsNullOrEmpty(status) || string.IsNullOrEmpty(priority))
            return SqliteRc.Misuse;
        if (title.Length is < 1 or > TitleMax || description.Length > DescMax)
            return SqliteRc.Range;
        if (status is not ("pending" or "in_progress" or "done" or "cancelled" or "archived"))
            return SqliteRc.Constraint;
        if (priority is not ("low" or "medium" or "high" or "urgent"))
            return SqliteRc.Constraint;
        return SqliteRc.Ok;
    }

    public static int StoreCategoriesForeach(TodoStore? s, Action<TodoStoreCategoryRow> fn)
    {
        if (s is null)
            return SqliteRc.Misuse;
        const string q = "SELECT id, name, color FROM categories ORDER BY name COLLATE NOCASE ASC;";
        try
        {
            using var cmd = s.Conn.CreateCommand();
            cmd.CommandText = q;
            using var reader = cmd.ExecuteReader();
            while (reader.Read())
            {
                fn(new TodoStoreCategoryRow(
                    reader.GetInt64(0),
                    reader.IsDBNull(1) ? "" : reader.GetString(1),
                    reader.IsDBNull(2) ? "" : reader.GetString(2)));
            }
            return SqliteRc.Ok;
        }
        catch (SqliteException)
        {
            return SqliteRc.Error;
        }
    }

    public static int StoreCategoryInsert(TodoStore? s, string name, string color, out long outId)
    {
        outId = 0;
        if (s is null)
            return SqliteRc.Misuse;
        var v = ValidateNameColor(name, color);
        if (v != SqliteRc.Ok)
            return v;
        const string q = "INSERT INTO categories(name, color) VALUES($name, $color);";
        try
        {
            using var cmd = s.Conn.CreateCommand();
            cmd.CommandText = q;
            cmd.Parameters.AddWithValue("$name", name);
            cmd.Parameters.AddWithValue("$color", color);
            cmd.ExecuteNonQuery();
            using var idCmd = s.Conn.CreateCommand();
            idCmd.CommandText = "SELECT last_insert_rowid();";
            outId = (long)(idCmd.ExecuteScalar() ?? 0L);
            return SqliteRc.Ok;
        }
        catch (SqliteException ex) when (ex.SqliteErrorCode == 19)
        {
            return SqliteRc.Constraint;
        }
        catch (SqliteException)
        {
            return SqliteRc.Error;
        }
    }

    public static int StoreCategoryDeleteById(TodoStore? s, long id)
    {
        if (s is null)
            return SqliteRc.Misuse;
        const string q = "DELETE FROM categories WHERE id = $id;";
        try
        {
            using var cmd = s.Conn.CreateCommand();
            cmd.CommandText = q;
            cmd.Parameters.AddWithValue("$id", id);
            return cmd.ExecuteNonQuery() > 0 ? SqliteRc.Ok : SqliteRc.Error;
        }
        catch (SqliteException)
        {
            return SqliteRc.Error;
        }
    }

    public static int StoreTodosForeach(TodoStore? s, Action<TodoStoreTodoRow> fn)
    {
        if (s is null)
            return SqliteRc.Misuse;
        const string q =
            "SELECT id, title, description, status, priority, category_id, parent_id, due_date," +
            " position, created_at, updated_at FROM todos ORDER BY position ASC, id ASC;";
        try
        {
            using var cmd = s.Conn.CreateCommand();
            cmd.CommandText = q;
            using var reader = cmd.ExecuteReader();
            while (reader.Read())
            {
                var catNull = reader.IsDBNull(5);
                var parNull = reader.IsDBNull(6);
                var dueNull = reader.IsDBNull(7);
                fn(new TodoStoreTodoRow(
                    reader.GetInt64(0),
                    reader.IsDBNull(1) ? "" : reader.GetString(1),
                    reader.IsDBNull(2) ? "" : reader.GetString(2),
                    reader.IsDBNull(3) ? "" : reader.GetString(3),
                    reader.IsDBNull(4) ? "" : reader.GetString(4),
                    catNull,
                    catNull ? 0 : reader.GetInt64(5),
                    parNull,
                    parNull ? 0 : reader.GetInt64(6),
                    dueNull,
                    dueNull ? 0 : reader.GetInt64(7),
                    reader.IsDBNull(8) ? 0 : reader.GetInt32(8),
                    reader.IsDBNull(9) ? 0 : reader.GetInt64(9),
                    reader.IsDBNull(10) ? 0 : reader.GetInt64(10)));
            }
            return SqliteRc.Ok;
        }
        catch (SqliteException)
        {
            return SqliteRc.Error;
        }
    }

    public static int StoreTodoInsert(
        TodoStore? s,
        string title,
        string description,
        string status,
        string priority,
        bool categoryIdSet,
        long categoryId,
        bool parentIdSet,
        long parentId,
        bool dueDateSet,
        long dueDate,
        out long outId)
    {
        outId = 0;
        if (s is null)
            return SqliteRc.Misuse;
        var v = ValidateTodoTexts(title, description, status, priority);
        if (v != SqliteRc.Ok)
            return v;
        const string q =
            "INSERT INTO todos(title, description, status, priority, category_id, parent_id, due_date, position)" +
            " VALUES($t,$d,$st,$pr,$cat,$par,$due, (SELECT COALESCE(MAX(position), -1) + 1 FROM todos));";
        try
        {
            using var cmd = s.Conn.CreateCommand();
            cmd.CommandText = q;
            cmd.Parameters.AddWithValue("$t", title);
            cmd.Parameters.AddWithValue("$d", description);
            cmd.Parameters.AddWithValue("$st", status);
            cmd.Parameters.AddWithValue("$pr", priority);
            cmd.Parameters.AddWithValue("$cat", categoryIdSet && categoryId > 0 ? categoryId : DBNull.Value);
            cmd.Parameters.AddWithValue("$par", parentIdSet && parentId > 0 ? parentId : DBNull.Value);
            cmd.Parameters.AddWithValue("$due", dueDateSet ? dueDate : DBNull.Value);
            cmd.ExecuteNonQuery();
            using var idCmd = s.Conn.CreateCommand();
            idCmd.CommandText = "SELECT last_insert_rowid();";
            outId = (long)(idCmd.ExecuteScalar() ?? 0L);
            return SqliteRc.Ok;
        }
        catch (SqliteException)
        {
            return SqliteRc.Error;
        }
    }

    public static int StoreTodoUpdateFull(
        TodoStore? s,
        long id,
        string title,
        string description,
        string status,
        string priority,
        bool categoryIdSet,
        long categoryId,
        bool parentIdSet,
        long parentId,
        bool dueDateSet,
        long dueDate,
        int position)
    {
        if (s is null)
            return SqliteRc.Misuse;
        var v = ValidateTodoTexts(title, description, status, priority);
        if (v != SqliteRc.Ok)
            return v;
        const string q =
            "UPDATE todos SET title=$t, description=$d, status=$st, priority=$pr, category_id=$cat, parent_id=$par," +
            " due_date=$due, position=$pos, updated_at=unixepoch() WHERE id=$id;";
        try
        {
            using var cmd = s.Conn.CreateCommand();
            cmd.CommandText = q;
            cmd.Parameters.AddWithValue("$t", title);
            cmd.Parameters.AddWithValue("$d", description);
            cmd.Parameters.AddWithValue("$st", status);
            cmd.Parameters.AddWithValue("$pr", priority);
            cmd.Parameters.AddWithValue("$cat", categoryIdSet && categoryId > 0 ? categoryId : DBNull.Value);
            cmd.Parameters.AddWithValue("$par", parentIdSet && parentId > 0 ? parentId : DBNull.Value);
            cmd.Parameters.AddWithValue("$due", dueDateSet ? dueDate : DBNull.Value);
            cmd.Parameters.AddWithValue("$pos", position);
            cmd.Parameters.AddWithValue("$id", id);
            return cmd.ExecuteNonQuery() > 0 ? SqliteRc.Ok : SqliteRc.Error;
        }
        catch (SqliteException)
        {
            return SqliteRc.Error;
        }
    }

    public static int StoreTodoDeleteById(TodoStore? s, long id)
    {
        if (s is null)
            return SqliteRc.Misuse;
        const string q = "DELETE FROM todos WHERE id = $id;";
        try
        {
            using var cmd = s.Conn.CreateCommand();
            cmd.CommandText = q;
            cmd.Parameters.AddWithValue("$id", id);
            return cmd.ExecuteNonQuery() > 0 ? SqliteRc.Ok : SqliteRc.Error;
        }
        catch (SqliteException)
        {
            return SqliteRc.Error;
        }
    }
}
